### DISCUSSION TITLE
Integrations API: support per-page metadata

### DISCUSSION THREAD

**Discussion by jamesarosen** [withastro/roadmap#515] [Proposal]
### Body

The [@astrojs/sitemap docs](https://docs.astro.build/en/guides/integrations-guide/sitemap/) say,

> Due to limitations of Astro’s [Integration API](https://docs.astro.build/en/reference/integrations-reference/), this integration can’t analyze a given page’s source code. This configuration option can set changefreq, lastmod and priority on a site-wide basis; see the next option serialize for how you can set these values on a per-page basis.

The way to configure per-page rules is to write a case statement or if/else chain that tests absolute URLs:

```
export default defineConfig({
  site: 'https://stargazers.club',
  integrations: [
    sitemap({
      serialize(item) {
        if (/exclude-from-sitemap/.test(item.url)) {
          return undefined;
        }
        if (/your-special-page/.test(item.url)) {
          item.changefreq = 'daily';
          item.lastmod = new Date();
          item.priority = 0.9;
        }
        return item;
      },
    }),
  ],
});
```

This isn't too bad for static rules like "Assign `priority=1.0` to the home page." But it's miserable to maintain for rules like "assign priority=0.7 to the most recent 10 blog posts and 0.4 to all others."

I would love to be able to do something like this:

```
// astro.config.ts
async serialize(item) {
  const module = await import(item.modulePath)
  return { ...item, ...module.rss }
}

// pages/index.astro
---
export const rss = { priority: 1.0 }
---
Hello, World
```

It's possible that the `rss` export depends on some Collection data or query-params. One way to make that work would be to have an earlier build step generate not only the HTML (or whatever) for the URL, but also an ambient module so that `item.modulePath` is actually something like `astro:pages/index.html` or `astro:pages/blog/my-slug.html`.

**Comment by Akxe** [comment]
I thought about adding a new export variable to pages that could enhance the sitemap generation.

```ts
---
import './project.scss';

interface Props {
  entry: CollectionEntry<'project'>;
}

export async function getStaticPaths() {
  const work = await getCollection('project', entry => !entry.data.hidden);
  return work.map(entry => ({
    params: { slug: entry.slug },
    props: { entry },
  }));
}

export const sitemapHints = {
  priority: 0.9,
  image: {
    url: data.img,
    title: data.imgAlt,
  },
};
---

<!-- HTML -->
```
